src/core/Akka.Streams/Implementation/IO/FileSubscriber.cs

```diff
@@ -105,113 +117,158 @@ protected override void PreStart()
-                    catch (Exception ex)
-                    {
-                        _log.Error(ex, "Tearing down FileSink({0}). File flush failed.", _f.FullName);
-                        CloseAndComplete(IOResult.Failed(_bytesWritten, ex));
-                        Context.Stop(Self);
-                    }
+                    if (!_autoFlush)
+                        RequestFlush();
                     return true;
+                default:
+                    return false;
             }
-
-            return false;
         }
 
-        /// <summary>
-        /// TBD
-        /// </summary>
-        protected override void PostStop()
+        private void RequestFlush()
         {
-            CloseAndComplete(IOResult.Success(_bytesWritten));
-            base.PostStop();
+            _flushRequested = true;
+            if (_fileRequests == 0)
+            {
+                Self.Tell(new Status.Success(ByteString.Empty));
+                _fileRequests++;
+            }
         }
 
-        private void CloseAndComplete(Try<IOResult> result)
+        private async Task WriteAsync(IActorRef self, ByteString byteString, bool flush)
         {
             try
             {
-                // close the channel/file before completing the promise, allowing the
-                // file to be deleted, which would not work (on some systems) if the
-                // file is still open for writing
-                _chan?.Dispose();
-
-                if (result.IsSuccess) 
-                    _completionPromise.SetResult(result.Success.Value);
-                else 
-                    _completionPromise.SetException(result.Failure.Value);
+                if (byteString.Count > 0)
+                    await byteString.WriteToAsync(_chan, _cts.Token).ConfigureAwait(false);
+
+                if (flush)
+                    await _chan.FlushAsync(_cts.Token).ConfigureAwait(false);
+
+                self.Tell(new Status.Success(byteString));
             }
             catch (Exception ex)
             {
-                _completionPromise.TrySetException(ex);
+                self.Tell(new Status.Failure(ex, "write_failed"));
             }
         }
 
+        /// <summary>
+        /// TBD
+        /// </summary>
+        protected override void PostStop()
+        {
+            _cts.Cancel();
+            _chan?.Dispose();
+            _completionPromise.TrySetCanceled();
+        }
+
         internal class FlushSignal
         {
             public static readonly FlushSignal Instance = new FlushSignal();
             private FlushSignal() { }
         }
+
+        sealed class FileStreamRequestStrategy : MaxInFlightRequestStrategy
+        {
+            public FileStreamRequestStrategy(int max) : base(max)
+            {
+            }
+
+            /// <summary>
+            /// How many complete file stream buffers are queued
+            /// </summary>
+            public int FileBufferCount { get; set; }
```

Comment:
`FileBufferCount` is used for the backpressure feature, it need to be mutable
default are 16 max, 5 min and file-buffer default 4KB 
Meaning that it will request until buffer is 64KB big